docs/source/reference/deprecation.rst

```diff
@@ -0,0 +1,327 @@
+===================
+Deprecation Notices
+===================
+
+This section contains information about deprecation of behaviours, features and
+APIs that have become undesirable/obsolete. Any information about the schedule
+for their deprecation and reasoning behind the changes, along with examples, is
+provided. However, first is a small section on how to suppress deprecation
+warnings that may be raised from Numba so as to prevent warnings propagating
+into code that is consuming Numba.
+
+Suppressing Deprecation warnings
+================================
+All Numba deprecations are issued via ``NumbaDeprecationWarning`` or
+``NumbaPendingDeprecationWarning`` s, to suppress the reporting of
+these the following code snippet can be used::
+
+    from numba.errors import NumbaDeprecationWarning, NumbaPendingDeprecationWarning
+    import warnings
+
+    warnings.simplefilter('ignore', category=NumbaDeprecationWarning)
+    warnings.simplefilter('ignore', category=NumbaPendingDeprecationWarning)
+
+The ``action`` used above is ``'ignore'``, other actions are available, see
+`The Warnings Filter <https://docs.python.org/3/library/warnings.html#the-warnings-filter>`_
+documentation for more information.
+
+Deprecation of reflection for List and Set types
+================================================
+Reflection (:term:`reflection`) is the jargon used in Numba to describe the
+process of ensuring that changes made by compiled code to arguments that are
+mutable Python container data types are visible in the Python interpreter when
+the compiled function returns. Numba has for some time supported reflection of
+``list`` and ``set`` data types and it is support for this reflection that
+is scheduled for deprecation with view to replace with a better implementation.
+
+Reason for deprecation
+----------------------
+First recall that for Numba to be able to compile a function in ``nopython``
+mode all the variables must have a concrete type ascertained through type
+inference. For functions which have mutable Python containers as arguments the
+type of these arguments is the container type, but the container type
+must also have a type, and then if the container type is also a container type
+it must also have a type and so on. Essentially it is relatively easy to build
+up containers with a high level of type complexity, and because the containers
+are arguments any updates made to them must be ``reflected`` back into the
+interpreter. After a number of years of experience with this problem, it is
```

Comment:
I see what you are going for, but everything between "For functions" and "back into the interpreter" is very confusing.  Maybe something more like this?:

```
In simple cases, it is clear how to reflect changes to containers inside nopython mode back to the original Python containers.  However, reflecting changes to complex data structures with nested container types (for example, lists of lists of integers) quickly becomes impossible to do efficiently and consistently.
```

Reply:
Thanks, I wasn't too happy with the original and this is a load better so I've replaced it.